File: Framework/LanguageSupport/python/MMIPython/src/MMIPython/abstraction/access/local/local_adapter_access.py
## SPDX-License-Identifier: MIT
## The content of this file has been developed in the context of the MOSIM research project.
## Original author(s): Jannes Lehwald

# -*- coding: utf-8 -*-
"""

"""

from MMIStandard.register.ttypes import MAdapterDescription
from MMIStandard.register import MMIAdapter

from MMIPython.abstraction.access.interface.adapter_access import IAdapterAccess
import MMIPython.abstraction.mmu.mmu_access
from MMIPython.abstraction.access.local.local_adapter_client import LocalAdapterClient

class LocalAdapterAccess(IAdapterAccess):
    
    """
    The adapter access.
    
    Attributes
    ----------
    description : AdapterDescription
        The description of the adapter
    initialized : bool
        Flag which indicates whether the communciaton has been successfully initialized
    aborted : bool
        Flag which indicates whether the initialization has been aborted
    scene_synchronized : bool
        Flag which indicates whether the scene has been successfully synchronized
    loaded : bool
        Flag which indicates whether the mmus of this adapter have been succesfully loaded
    mmu_descriptions : list<MMUDescription>
        The list of mmus of this adpater
    
    _client : IAdapterClient
        The client adapter
    _mmuAccess : MMUAccess
        The MMU Access
    _instance : MMIAdapter.Iface
        The local adapter
    """
    
    def __init__(self, description, instance, mmuAccess):
        
        """
        Basic initialization of an adapter access
        
        Parameters
        ----------
        description : AdapterDescription
            The description of the adapter
        instance : MMIAdapter.Iface
            The port of the adapter
        mmuAccess : MMUAccess
            The MMU Access
        """
        
        assert (isinstance(mmuAccess, MMIPython.abstraction.mmu.mmu_access.MMUAccess)),"mmuAccess is no MMUAccess"
        assert (isinstance(description, MAdapterDescription)),"description is no AdapterDescription"
        assert (isinstance(instance, MMIAdapter.Iface)), "instance is no MMIAdapter.Iface"
        
        super(LocalAdapterAccess, self).__init__(mmuAccess)
        
        self.description = description
        self._instance = instance
        
    def start(self):
        """
        Method starts the process
        """
        self._client = self.create_client()
        self.initialized = True
        
    def create_client(self):
        """
        Creates a client for communication with the adapter
        
        Returns
        -------
        IAdapterClient
            The communication client
        """
        return LocalAdapterClient(self._instance)

File: Framework/LanguageSupport/cs/MMICSharp/Access/Abstraction/LocalAdapterAccess.cs
// SPDX-License-Identifier: MIT
// The content of this file has been developed in the context of the MOSIM research project.
// Original author(s): Felix Gaisbauer

using System;
using System.Collections.Generic;
using MMIStandard;


namespace MMICSharp.Access.Abstraction
{
    /// <summary>
    /// An adapter access which can directly to connect to a local CSharp adapter
    /// </summary>
    public class LocalAdapterAccess : IAdapter
    {
        #region public properties
        public MAdapterDescription Description
        {
            get;
            set;
        }

        public bool Initialized
        {
            get;
            set;
        }

        public bool SceneSynchronized
        {
            get;
            set;
        }

        public List<MMUDescription> MMUDescriptions
        {
            get;
            set;
        } = new List<MMUDescription>();


        public bool Loaded
        {
            get;
            set;
        } = false;

        #endregion


        private class LocalAdapterClient : IAdapterClient
        {
            public MMIAdapter.Iface Access
            {
                get;
                set;
            }

            public LocalAdapterClient(MMIAdapter.Iface instance)
            {
                this.Access = instance;
            }

            public void Dispose()
            {
                //Nothing to do
            }
        }

        private readonly MMIAdapter.Iface instance;
        private readonly MMUAccess mmuAccess;

        /// <summary>
        /// Basic constructor
        /// </summary>
        /// <param name="description"></param>
        /// <param name="instance"></param>
        public LocalAdapterAccess(MAdapterDescription description, MMIAdapter.Iface instance, MMUAccess mmuAccess)
        {
            this.Description = description;
            this.instance = instance;
            this.mmuAccess = mmuAccess;
        }


        public void Start()
        {
            this.Initialized = true;
        }

        public IAdapterClient CreateClient()
        {
            return new LocalAdapterClient(this.instance);
        }

        public byte[] CreateCheckpoint(string mmuID, string checkpointID)
        {
            return instance.CreateCheckpoint(mmuID, mmuAccess.SessionId);
        }

        public List<MotionModelUnitAccess> CreateMMUConnections(string sessionId, List<MMUDescription> mmuDescriptions)
        {
            List<MMUDescription> availableMMUs = this.instance.GetMMus(sessionId);

            if (availableMMUs == null)
                throw new Exception("No MMUs available");



            List<MotionModelUnitAccess> result = new List<MotionModelUnitAccess>();
            foreach (MMUDescription description in availableMMUs)
            {
                //Create a new MMMU connection instance
                result.Add(new MotionModelUnitAccess(this.mmuAccess, this, this.mmuAccess.SessionId, description));
            }

            return result;
        }

        public MBoolResponse CreateSession(string sessionId, MAvatarDescription referenceAvatar)
        {
            return this.instance.CreateSession(sessionId);
        }

        public MBoolResponse Dispose()
        {
            throw new NotImplementedException();
        }

        public List<MMUDescription> GetLoadableMMUs(string sessionId)
        {
            this.MMUDescriptions = this.instance.GetLoadableMMUs();
            return this.MMUDescriptions;
        }

        public List<MSceneObject> GetScene(string sessionId)
        {
            return this.instance.GetScene(sessionId);
        }

        public MSceneUpdate GetSceneChanges(string sessionId)
        {
            return this.instance.GetSceneChanges(sessionId);
        }

        public Dictionary<string, string> GetStatus()
        {
            return this.instance.GetStatus();
        }

        public MBoolResponse LoadMMUs(List<string> ids, string sessionId)
        {
            this.instance.LoadMMUs(ids, sessionId);

            this.Loaded = true;

            return new MBoolResponse(true);
        }

        public MBoolResponse RestoreCheckpoint(string mmuID, string checkpointID,byte[] checkpointData)
        {
            return this.instance.RestoreCheckpoint(mmuID, this.mmuAccess.SessionId, checkpointData);
        }



        public MBoolResponse PushScene(MSceneUpdate sceneUpdates, string sessionId)
        {
            return this.instance.PushScene(sceneUpdates, sessionId);
        }

        public MBoolResponse CloseSession(string sessionID)
        {
            return this.instance.CloseSession(sessionID);
        }

        public MBoolResponse CloseConnection()
        {
            //Nothing to do -> since local access

            return new MBoolResponse(true);
        }
    }
}
